pallets/router/src/lib.rs

```diff
@@ -0,0 +1,193 @@
+// Copyright 2021 Parallel Finance Developer.
+// This file is part of Parallel Finance.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! # Route for Automatic Market Maker (AMM)
+//!
+//! Given a supported `route` like this `(0, USDT, KSM)`, we can get corresponding AMM or pool.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use crate::pallet::*;
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+#[frame_support::pallet]
+pub mod pallet {
+    use frame_support::{
+        ensure,
+        pallet_prelude::DispatchResultWithPostInfo,
+        traits::{Get, Hooks, IsType},
+        transactional, PalletId,
+    };
+    use frame_system::{
+        ensure_signed,
+        pallet_prelude::{BlockNumberFor, OriginFor},
+    };
+    use orml_traits::{MultiCurrency, MultiCurrencyExtended};
+    use primitives::{Amount, Balance, CurrencyId, AMM};
+    use sp_runtime::traits::Zero;
+
+    pub type Route = Vec<(
+        // ID of the AMM to use, as specified in the `Config` trait. Setting this
+        // to 0 would take the first AMM instance specified in `type AMMs`.
+        u8,
+        // Base asset
+        CurrencyId,
+        // Quote asset
+        CurrencyId,
+    )>;
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+        /// Route pallet id
+        #[pallet::constant]
+        type RoutePalletId: Get<PalletId>;
+
+        /// Specify all the AMMs we are routing between
+        type AMMAdaptor: AMM<Self>;
```

Comment:
```suggestion
        type AMM: AMM<Self>;
```

only one AMM and no adapter in this simplified design